services/datamanager/builtin/sync/upload_arbitrary_file.go

```diff
@@ -0,0 +1,156 @@
+package sync
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/pkg/errors"
+	v1 "go.viam.com/api/app/datasync/v1"
+
+	"go.viam.com/rdk/logging"
+)
+
+// UploadChunkSize defines the size of the data included in each message of a FileUpload stream.
+var (
+	UploadChunkSize            = 64 * 1024
+	errFileEmpty               = errors.New("file is empty (0 bytes)")
+	errFileModifiedTooRecently = errors.New("file modified too recently")
+)
+
+func uploadArbitraryFile(
+	ctx context.Context,
+	f *os.File,
+	conn cloudConn,
+	tags []string,
+	fileLastModifiedMillis int,
+	clock clock.Clock,
+	logger logging.Logger,
+) error {
+	logger.Debugf("attempting to sync arbitrary file: %s", f.Name())
+	path, err := filepath.Abs(f.Name())
+	if err != nil {
+		return errors.Wrap(err, "failed to get absolute path")
+	}
+
+	// Only sync non-datacapture files that have not been modified in the last
+	// fileLastModifiedMillis to avoid uploading files that are being
+	// to written to.
+	info, err := os.Stat(path)
+	if err != nil {
+		return errors.Wrap(err, "stat failed")
+	}
+	if info.Size() == 0 {
+		return errFileEmpty
+	}
+
+	timeSinceMod := clock.Since(info.ModTime())
+	if timeSinceMod < time.Duration(fileLastModifiedMillis)*time.Millisecond {
+		return errFileModifiedTooRecently
+	}
+
+	// We need to seek to the start of the file as if we have tried to upload this file
+	// previously, the read offset of the file might not be at the beginning, which would
+	// result in partial data loss when uploading to the cloud.
+	// Fixes https://viam.atlassian.net/browse/DATA-3114
+	pos, err := f.Seek(0, io.SeekStart)
+	if err != nil {
+		return errors.Wrap(err, "error trying to Seek to beginning of file")
+	}
+
+	if pos != 0 {
+		msg := "error trying to Seek to beginning of file %s attempted to seek to " +
+			"beginning of file but only got to position: %d, expected to be at position 0"
+		return fmt.Errorf(msg, path, pos)
```

Comment:
nit, why is `msg` not inlined here? also the text is a little repetitive, could be something like `error trying to seek to beginning of file %s: expected position 0, instead got to position %d.`

Reply:
done